src/tests/suite/daos_nvme_recovery.c

```diff
@@ -254,11 +254,203 @@ nvme_test_verify_device_stats(void **state)
 	D_FREE(devices);
 }
 
+/* Verify NVMe I/O error and notification*/
+static void
+nvme_test_simulate_IO_error(void **state)
+{
+	test_arg_t	*arg = *state;
+	device_list	*devices = NULL;
+	daos_obj_id_t	oid;
+	struct ioreq	req;
+	const char	dkey[] = "dkey";
+	const char	akey[] = "akey";
+	daos_size_t	size = 4 * 4096; /* record size */
+	char		*ow_buf;
+	char		*fbuf;
+	char		*write_errors;
+	char		*read_errors;
+	char		*check_errors;
+	char		*control_log_file;
+	char		*server_config_file;
+	int		rx_nr; /* number of record extents */
+	int		rank_pos = 0, rank = 1;
+	int		ndisks, rc, i;
+
+	if (!is_nvme_enabled(arg)) {
+		print_message("NVMe isn't enabled.\n");
+		skip();
+	}
+
+	/*
+	 *  Get the Total number of NVMe devices from all the servers.
+	 */
+	rc = dmg_storage_device_list(dmg_config_file, &ndisks, NULL);
+	assert_int_equal(rc, 0);
+	print_message("Total Disks = %d\n", ndisks);
+
+	/*
+	 * Get the Device info of all NVMe devices
+	 */
+	D_ALLOC_ARRAY(devices, ndisks);
+	rc = dmg_storage_device_list(dmg_config_file, NULL, devices);
+	assert_int_equal(rc, 0);
+
+	/*
+	 * Get the rank 1 position from array devices
+	 */
+	for (i = 0; i < ndisks; i++) {
+		if (devices[i].rank == 1) {
+			rank_pos = i;
+			break;
+		}
+	}
+
+	/*
+	 * Get DAOS server file
+	 */
+	D_ALLOC(control_log_file, 1024);
+	D_ALLOC(server_config_file, 512);
+	rc = get_server_config(devices[rank_pos].host, server_config_file);
+	assert_int_equal(rc, 0);
+	print_message("server_config_file = %s\n", server_config_file);
+
+	/*
+	 * Get DAOS control log file
+	 */
+	get_log_file(devices[rank_pos].host, server_config_file,
+		     "control_log_file", control_log_file);
+	print_message("Control Log File = %s\n", control_log_file);
+	D_FREE(server_config_file);
+
+	/*
+	 * Get the Initial write error
+	 */
+	write_errors = strdup("bio_write_errs");
+	rc = dmg_storage_query_device_health(dmg_config_file,
+					     devices[rank_pos].host,
+					     write_errors,
+		devices[rank_pos].device_id);
+	assert_int_equal(rc, 0);
+	print_message("Initial write_errors = %s\n", write_errors);
+
+	/*
+	 * Get the Initial read error
+	 */
+	read_errors = strdup("bio_read_errs");
+	rc = dmg_storage_query_device_health(dmg_config_file,
+					     devices[rank_pos].host,
+					     read_errors,
+		devices[rank_pos].device_id);
+	assert_int_equal(rc, 0);
+	print_message("Initial read_errors = %s\n", read_errors);
+
+	/*
+	 * Allocate and set write buffer with data
+	 */
+	D_ALLOC(ow_buf, size);
+	assert_non_null(ow_buf);
+	dts_buf_render(ow_buf, size);
+	/*
+	 * Allocate and set fetch buffer
+	 */
+	D_ALLOC(fbuf, size);
+	assert_non_null(fbuf);
+	memset(fbuf, 0, size);
+
+	/*
+	 * Inject BIO Write Errors on Rank1 device
+	 */
+	print_message("----Inject BIO Write Error----\n");
+	set_fail_loc(arg, rank, DAOS_NVME_BIO_WRITE_ERR | DAOS_FAIL_ONCE);
```

Comment:
why inject write errors before any data is written?

Reply:
This will be only set and and not inject. Later on Write when rw_completion will be called and that will inject the error. I put the delay after set_fail_loc to see when that is getting called and it occurs only when Write happens.
Let me know if I am still missing the flow.